In [1]:
import pandas as pd
df3 = pd.read_csv('df2.csv')

In [2]:
df3.head()

depdelay  arrdelay  nonhubairportorigin  smallhubairportorigin  \
0         0      -4.0                    1                      0   
1        -4      11.0                    0                      0   
2         3      12.0                    0                      0   
3        -3      24.0                    0                      0   
4         0      -8.0                    0                      0   

   mediumhubairportorigin  largehubairportorigin  nonhubairportdest  \
0                       0                      0                  1   
1                       0                      1                  0   
2                       0                      1                  0   
3                       1                      0                  0   
4                       0                      1                  1   

   smallhubairportdest  mediumhubairportdest  largehubairportdest  ...  \
0                    0                     0                    0  ...   
1                    1                     0                    0  ...   
2                    0                     1                    0  ...   
3                    0                     0                    1  ...   
4                    0                     0                    0  ...   

   temp_40_infty  windspeed  windspeedsquare  windgustdummy  windgustspeed  \
0              0  10.300000       106.090000              0            0.0   
1              0  13.666667       186.777778              0            0.0   
2              0  13.528302       183.014952              0            0.0   
3              0   0.000000         0.000000              0            0.0   
4              0   7.666667        58.777778              0            0.0   

   raindummy  raintracedummy  snowdummy  snowtracedummy  delay_sum  
0          0               0          0               0       -4.0  
1          0               0          0               0        7.0  
2          0               0          0               0       15.0  
3          1               0          0               0       21.0  
4          0               0          0               0       -8.0  

[5 rows x 47 columns]

In [10]:
import numpy as np
df3['delayed'] = np.where(df3['delay_sum']>15, 1, 0) #created a new column such that 1 indicates a delayed flight, 0 indicates an on-time flight
df3.head()

depdelay  arrdelay  nonhubairportorigin  smallhubairportorigin  \
0         0      -4.0                    1                      0   
1        -4      11.0                    0                      0   
2         3      12.0                    0                      0   
3        -3      24.0                    0                      0   
4         0      -8.0                    0                      0   

   mediumhubairportorigin  largehubairportorigin  nonhubairportdest  \
0                       0                      0                  1   
1                       0                      1                  0   
2                       0                      1                  0   
3                       1                      0                  0   
4                       0                      1                  1   

   smallhubairportdest  mediumhubairportdest  largehubairportdest  ...  \
0                    0                     0                    0  ...   
1                    1                     0                    0  ...   
2                    0                     1                    0  ...   
3                    0                     0                    1  ...   
4                    0                     0                    0  ...   

   windspeed  windspeedsquare  windgustdummy  windgustspeed  raindummy  \
0  10.300000       106.090000              0            0.0          0   
1  13.666667       186.777778              0            0.0          0   
2  13.528302       183.014952              0            0.0          0   
3   0.000000         0.000000              0            0.0          1   
4   7.666667        58.777778              0            0.0          0   

   raintracedummy  snowdummy  snowtracedummy  delay_sum  delayed  
0               0          0               0       -4.0        0  
1               0          0               0        7.0        0  
2               0          0               0       15.0        0  
3               0          0               0       21.0        1  
4               0          0               0       -8.0        0  

[5 rows x 48 columns]

In [11]:

import pandas as pd
from sklearn.preprocessing import StandardScaler
 
 
# Initialise the Scaler
scaler = StandardScaler()
 
# To scale data
scaler.fit(df3)

StandardScaler()

In [12]:
>>> import pandas as pd
>>> from sklearn.model_selection import train_test_split
>>> from sklearn.datasets import load_iris

y = df3.delayed
x = df3.drop('delayed', axis = 1)

In [13]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2)
x_train.head()


depdelay  arrdelay  nonhubairportorigin  smallhubairportorigin  \
514598         25      15.0                    0                      0   
575583         -1      -9.0                    0                      0   
1190216        -1     -19.0                    1                      0   
79253          -3     -11.0                    0                      0   
1003119        -5      -4.0                    0                      1   

         mediumhubairportorigin  largehubairportorigin  nonhubairportdest  \
514598                        1                      0                  0   
575583                        1                      0                  0   
1190216                       0                      0                  0   
79253                         0                      1                  0   
1003119                       0                      0                  0   

         smallhubairportdest  mediumhubairportdest  largehubairportdest  ...  \
514598                     0                     1                    0  ...   
575583                     0                     0                    1  ...   
1190216                    0                     1                    0  ...   
79253                      1                     0                    0  ...   
1003119                    0                     0                    1  ...   

         temp_40_infty  windspeed  windspeedsquare  windgustdummy  \
514598               0      11.00         121.0000              1   
575583               0       4.75          22.5625              0   
1190216              0       0.50           0.2500              0   
79253                0       5.15          26.5225              0   
1003119              0       6.00          36.0000              0   

         windgustspeed  raindummy  raintracedummy  snowdummy  snowtracedummy  \
514598        6.792453          0               0          0               0   
575583        0.000000          0               0          0               0   
1190216       0.000000          0               0          0               0   
79253         0.000000          1               0          0               0   
1003119       0.000000          0               0          0               0   

         delay_sum  
514598        40.0  
575583       -10.0  
1190216      -20.0  
79253        -14.0  
1003119       -9.0  

[5 rows x 47 columns]

In [14]:
y_train.head()


514598     1
575583     0
1190216    0
79253      0
1003119    0
Name: delayed, dtype: int32

## Now that I've separated the dataset into my training sets and test sets, I will decide which model is best. There are two categories of models: classification and regression. Classification is for discrete labels, but not continuous values (e.g. 0 or 1, yes/no); regression is for continuous values. In my case, I want to find a discrete label (i.e. 1 for a delayed flight, and 0 for an on-time flight). As such, I will explore different classification metrics.

In [50]:
from sklearn.metrics import precision_score, recall_score, f1_score
print("Precision: ", precision_score(y_test, model.predict(x_test)))
print('recall: ', recall_score(y_test, model.predict(x_test)))
print("F1 score: ", f1_score(y_test, model.predict(x_test)))

Precision:  0.8181818181818182
recall:  0.3879310344827586
F1 score:  0.5263157894736843


In [64]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size = 0.2, random_state = 42)
LR_model = LogisticRegression().fit(x_train,y_train)
LR_model.score(x_test,y_test)

0.9595

In [65]:
from sklearn.ensemble import RandomForestClassifier
RFC_model = RandomForestClassifier(random_state = 42).fit(x_train,y_train)
RFC_model.score(x_test,y_test)

0.964

#### From these three models, the Logistic Regression and Random Forest Classifier do the best, at 95.95% and 96.4%, respectively; however, it is worth noting that the latter took a notably longer time than the previous.